RFC: Asset definition decorator

clairelin135 opened the pull request:
This PR introduces an `@assets_definition` decorator that decorates an op to create an `AssetsDefinition`:

Code snippet from https://github.com/dagster-io/dagster/pull/7432:
```
@assets_definition
@op(out={"a": Out(), "b": Out()})
def multi_asset_op(x):
    return ...

# or, if you want to opt out of magic asset key guessing:

@assets_definition(asset_key_by_output_name={"a": AssetKey("a_asset"), "b": AssetKey("b_asset")})
@op(out={"a": Out(), "b": Out()})
def multi_asset_op(x):
    return ...
```
Specifying no arguments on `assets_definition` and `op` will automatically infer the input and output asset keys from the decorated function. For the below function:
- input asset keys would be `AssetKey("a")` and `AssetKey("b")`
- output asset key would be `AssetKey("result")` since by default the output of an op is named "result"
```
@assets_definition
@op
def multi_asset_op(a, b):
    return y
```

Some distinctions:
- `non_argument_deps` is no longer an argument to `assets_definition`. Because `assets_definition` decorates an op, any input with type `Nothing` will have asset key `AssetKey(input_def.name)`.
- `resource_defs` will not be an argument as all operational domain arguments will exist on the op definition layer
- Will probably need to add a description field on this decorator in the future, but punting on this for now

OwenKephart commented on python_modules/dagster/dagster/core/asset_defs/decorators.py:
```diff
@@ -301,6 +303,131 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+    The name of the argument designates the name of the upstream asset.
+
+    Args:
+        name (Optional[str]): The name of the op.
```
these args will need to be updated

OwenKephart commented on python_modules/dagster/dagster/core/asset_defs/decorators.py:
```diff
@@ -301,6 +303,131 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+    asset_key_by_output_name: Optional[Mapping[str, AssetKey]] = None,
+    internal_asset_deps: Optional[Mapping[str, Set[AssetKey]]] = None,
+) -> Callable[[Callable[..., Any]], AssetsDefinition]:
+    """Create a combined definition of multiple assets that are computed using the same op and same
```
Nitpick, but I might change the wording here a bit to something like "Create a definition of multiple assets which share the same computation and same upstream assets", just to make it a little more future-proof (for when graphs can be annotated as well)

OwenKephart reviewed (changes requested):
Overall, this looks basically perfect. Beyond the comments I had inline, I think this PR should replace the current implementations of the @asset and @multi_asset decorators to use a combination of @op and @assets_definition. They'd keep the same arguments and everything, but that way, we can eliminate a bunch of redundant code.

We might even want to slap a deprecation warning on multi_asset, but I'm impartial on that at the moment.

sryza commented on python_modules/dagster/dagster/core/asset_defs/decorators.py:
```diff
@@ -301,6 +303,131 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+                "to any key provided in the out dictionary of the decorated op"
+            )
+
+    for key in op_outs:
```
nitpick: would rename `key` to something like `output_name` for clarity

sryza commented on python_modules/dagster/dagster/core/asset_defs/decorators.py:
```diff
@@ -301,6 +303,131 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+    return inferred_asset_key_by_input_name
+
+
+def _infer_asset_key_by_output_name(op_def, asset_key_by_output_name):
```
Would it make sense to consolidate this implementation with `build_asset_outs`?

clairelin135 replied:
Added the input type annotations. I think that it makes sense to keep this function separate from `build_asset_outs`, `build_asset_outs` builds an `outs` dictionary and does checks with `internal_asset_deps` that won't apply to this new decorator

sryza reviewed (changes requested):
Mostly looks great! Had some comments on naming.

OwenKephart commented on python_modules/dagster/dagster/core/asset_defs/decorators.py:
```diff
@@ -301,6 +303,131 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+@experimental_decorator
+def assets_definition(
+    name: Optional[str] = None,
+    asset_key_by_input_name: Optional[Mapping[str, AssetKey]] = None,
```
this one is my fault, as this is based off of my branch (was going to sneakily fix this in both once I merged them lol)

clairelin135 commented:
@OwenKephart @sryza Thanks for the feedback, I've addressed your comments from above. The only open question that I think is remaining here is the name of this decorator. We can discuss this at the next lane meeting when this batch of PRs is ready to merge.

OwenKephart commented on python_modules/dagster/dagster/core/asset_defs/decorators.py:
```diff
@@ -309,6 +313,154 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+    # Infer non-argument deps for inputs with type In(nothing) with AssetKey(input_name)
+
+    input_param_names = []
+    if isinstance(op_def.compute_fn, DecoratedSolidFunction):
```
in what scenarios would this condition be false?

clairelin135 replied:
Invoking the decorated op when the op is not constructed using the @op decorator. I can add raise an error when this occurs

OwenKephart commented on python_modules/dagster/dagster/core/asset_defs/decorators.py:
```diff
@@ -309,6 +313,154 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+    op_def: OpDefinition, asset_keys_by_output_name: Mapping[str, AssetKey]
+) -> Mapping[AssetKey, str]:
+    inferred_output_name_by_asset_key: Dict[AssetKey, str] = {
+        asset_key: output_name for output_name, asset_key in asset_keys_by_output_name.items()
```
I think for the case that there's only a single output named "result", we should use the name of the op as the output key rather than the name of the output.

OwenKephart reviewed (commented):
I still think we should try to implement both @asset and @multi_asset as a composition of this new decorator and @op.

clairelin135 commented:
Closing this in favor of https://github.com/dagster-io/dagster/pull/7524. This new PR contains the implementation for decorating graphs.
